### PR TITLE
En 8398 tests transactions fees cross

### PR DESCRIPTION
**Pull request by miiu96**
Tests:  
- added a lot o new integration tests. The new tests verify if the cross-shard transactions are executed as expected on the source shard and the destination shard.

Bug fix: 
- fix move balance should not accumulate fees in cross-shard

### CODE REVIEW

File: integrationTests/vm/txsFee/builtInFunctions_test.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package txsFee
 
 import (
 	"encoding/hex"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
```
Comment (sasurobert) [inline integrationTests/vm/txsFee/builtInFunctions_test.go:5]: goimports

File: integrationTests/vm/txsFee/relayedMoveBalance_test.go [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package txsFee
 
 import (
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
```
Comment (sasurobert) [inline integrationTests/vm/txsFee/relayedMoveBalance_test.go:4]: goimports

File: integrationTests/vm/txsFee/relayedScCalls_test.go [hunk 3]
```diff
@@ -2,6 +2,7 @@ package txsFee
 
 import (
 	"encoding/hex"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
```
Comment (sasurobert) [inline integrationTests/vm/txsFee/relayedScCalls_test.go:5]: goimports

File: integrationTests/vm/txsFee/relayedScDeploy_test.go [hunk 4]
```diff
@@ -2,6 +2,7 @@ package txsFee
 
 import (
 	"errors"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
```
Comment (sasurobert) [inline integrationTests/vm/txsFee/relayedScDeploy_test.go:5]: goimports

File: integrationTests/vm/txsFee/scCalls_test.go [hunk 5]
```diff
@@ -3,6 +3,7 @@ package txsFee
 import (
 	"encoding/hex"
 	"fmt"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
```
Comment (sasurobert) [inline integrationTests/vm/txsFee/scCalls_test.go:6]: goimports

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: integrationTests/vm/txsFee/multiShard/builtInFunctions_test.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+// Test scenario
+// 1. Do a SC deploy on shard 1
+// 2. Do a ChangeOwnerAddress (owner of the deployed contract will be moved in shard 0)
+// 3. Do a ClaimDeveloperReward (cross shard call , the transaction will be executed on the source shard and the destination shard)
+// 4. Execute SCR from context destination on context source ( the new owner will receive the developer rewards)
+func TestBuiltInFunctionExecuteOnSourceAndDestinationShouldWork(t *testing.T) {
+	testContextSource := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 0)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/builtInFunctions_test.go:21]: t.Parallel() ?

File: integrationTests/vm/txsFee/multiShard/moveBalance_test.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+	expectedBalanceReceiver := big.NewInt(0)
+	utils.TestAccount(t, testContext.Accounts, scAddrBytes, 0, expectedBalanceReceiver)
+
+	// check accumulated fees
+	accumulatedFees := testContext.TxFeeHandler.GetAccumulatedFees()
+	require.Equal(t, big.NewInt(910), accumulatedFees)
+}
+
+func TestMoveBalanceExecuteOneSourceAndDestinationShard(t *testing.T) {
+	testContextSource := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 0)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/moveBalance_test.go:128]: t.Parallel() ?

File: integrationTests/vm/txsFee/multiShard/moveBalance_test.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go/core/vmcommon"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMoveBalanceShouldWork(t *testing.T) {
+	testContext := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 1)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/moveBalance_test.go:16]: t.Parallel() ?

File: integrationTests/vm/txsFee/multiShard/relayedBuiltInFunctions_test.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/vmcommon"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRelayedBuiltInFunctionExecuteOnRelayerAndDstShardShouldWork(t *testing.T) {
+	testContextRelayer := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 2)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/relayedBuiltInFunctions_test.go:16]: t.Parallel() ?

File: integrationTests/vm/txsFee/multiShard/relayedMoveBalance_test.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go/core/vmcommon"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRelayedMoveBalanceRelayerShard0InnerTxSenderAndReceiverShard1ShouldWork(t *testing.T) {
+	testContext := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 1)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/relayedMoveBalance_test.go:15]: t.Parallel() ?

File: integrationTests/vm/txsFee/multiShard/relayedMoveBalance_test.go [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+	// check balance inner tx receiver
+	utils.TestAccount(t, testContext.Accounts, rcvAddr, 0, big.NewInt(100))
+
+	// check accumulated fees
+	accumulatedFees := testContext.TxFeeHandler.GetAccumulatedFees()
+	require.Equal(t, big.NewInt(1000), accumulatedFees)
+}
+
+func TestRelayedMoveBalanceRelayerAndInnerTxSenderShard0ReceiverShard1(t *testing.T) {
+	testContext := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 1)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/relayedMoveBalance_test.go:58]: t.Parallel() ?

File: integrationTests/vm/txsFee/builtInFunctions_test.go (hunk 1 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/builtInFunctions_test.go:5]: done.

File: integrationTests/vm/txsFee/relayedMoveBalance_test.go (hunk 2 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/relayedMoveBalance_test.go]: done.

File: integrationTests/vm/txsFee/relayedScCalls_test.go (hunk 3 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/relayedScCalls_test.go]: done.

File: integrationTests/vm/txsFee/relayedScDeploy_test.go (hunk 4 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/relayedScDeploy_test.go]: done.

File: integrationTests/vm/txsFee/scCalls_test.go (hunk 5 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/scCalls_test.go]: done.

File: integrationTests/vm/txsFee/multiShard/relayedMoveBalance_test.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+	// check balance inner tx receiver
+	utils.TestAccount(t, testContext.Accounts, rcvAddr, 0, big.NewInt(100))
+
+	// check accumulated fees
+	accumulatedFees := testContext.TxFeeHandler.GetAccumulatedFees()
+	require.Equal(t, big.NewInt(1000), accumulatedFees)
+}
+
+func TestRelayedMoveBalanceRelayerAndInnerTxSenderShard0ReceiverShard1(t *testing.T) {
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/relayedMoveBalance_test.go:57]: Do we also have a test to cover the situation?TestRelayedMoveBalanceRelayerAndInnerTxReceiverShard0SenderShard1

File: integrationTests/vm/txsFee/multiShard/relayedScDeploy_test.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+	"github.com/ElrondNetwork/elrond-go/core/vmcommon"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/arwen"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRelayedSCDeployShouldWork(t *testing.T) {
+	testContextRelayer := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 2)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/relayedScDeploy_test.go:15]: t.Parallel() ?

File: integrationTests/vm/txsFee/multiShard/relayedTxScCalls_test.go [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+// inner transaction sender - shard 0
+// inner transaction receiver (smart contract address) - shard 1
+
+// 1. Do a SC deploy on shard 1
+// 2. Execute relayed transaction on shard 2
+// 3. Execute relayed transaction on shard 0
+// 4. Execute SCR with the smart contract call on shard 1
+// 5. Execute SCR with refund on relayer shard (shard 2)
+func TestRelayedTxScCallMultiShardShouldWork(t *testing.T) {
+	testContextRelayer := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 2)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/relayedTxScCalls_test.go:24]: t.Parallel() ?

File: integrationTests/vm/txsFee/multiShard/relayedTxScCalls_test.go [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+	// check accumulated fees
+	accumulatedFees = testContextRelayer.TxFeeHandler.GetAccumulatedFees()
+	require.Equal(t, big.NewInt(1870), accumulatedFees)
+
+	developerFees = testContextRelayer.TxFeeHandler.GetDeveloperFees()
+	require.Equal(t, big.NewInt(0), developerFees)
+}
+
+func TestRelayedTxScCallMultiShardFailOnInnerTxDst(t *testing.T) {
+	testContextRelayer := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 2)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/relayedTxScCalls_test.go:126]: t.Parallel() ?

File: integrationTests/vm/txsFee/multiShard/relayedTxScCalls_test.go [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+	testContextInnerSource := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 0)
+	defer testContextInnerSource.Close()
+
+	testContextInnerDst := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 1)
+	defer testContextInnerDst.Close()
+
+	pathToContract := "../../arwen/testdata/counter/output/counter.wasm"
+	scAddr, owner := utils.DoDeploy(t, testContextInnerDst, pathToContract)
+	testContextInnerDst.TxFeeHandler.CreateBlockStarted()
+	utils.CleaAccumulatedIntermediateTransactions(t, testContextInnerDst)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/relayedTxScCalls_test.go:138]: typo: Clea instead Clean ?

File: integrationTests/vm/txsFee/multiShard/scCalls_test.go [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go/core/vmcommon"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/txsFee/utils"
+	"github.com/stretchr/testify/require"
+)
+
+func TestScCallExecuteOnSourceAndDstShardShouldWork(t *testing.T) {
+	testContextSource := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 0)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/scCalls_test.go:15]: t.Parallel() ?

File: integrationTests/vm/txsFee/multiShard/scCalls_test.go [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+	expectedBalance = big.NewInt(6140)
+	vm.TestAccount(t, testContextSource.Accounts, sndAddr, 1, expectedBalance)
+
+	// check accumulated fees
+	accumulatedFees = testContextSource.TxFeeHandler.GetAccumulatedFees()
+	require.Equal(t, big.NewInt(100), accumulatedFees)
+}
+
+func TestScCallExecuteOnSourceAndDstShardInvalidOnDst(t *testing.T) {
+	testContextSource := vm.CreatePreparedTxProcessorWithVMsMultiShard(t, 0)
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/multiShard/scCalls_test.go:90]: t.Parallel() ?

File: integrationTests/vm/txsFee/utils/utils.go [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+	require.Nil(t, testContext.GetLatestError())
+
+	_, err = testContext.Accounts.Commit()
+	require.Nil(t, err)
+
+	expectedBalance := big.NewInt(89030)
+	vm.TestAccount(t, testContext.Accounts, owner, senderNonce+1, expectedBalance)
+
+	// check accumulated fees
+	accumulatedFee := testContext.TxFeeHandler.GetAccumulatedFees()
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/utils/utils.go:51]: accumulatedFees

File: integrationTests/vm/txsFee/utils/utils.go [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+func GetIntermediateTransactions(t *testing.T, testContext *vm.VMTestContext) []data.TransactionHandler {
+	scForwarder := testContext.ScForwarder
+	mockIntermediate, ok := scForwarder.(*mock.IntermediateTransactionHandlerMock)
+	require.True(t, ok)
+
+	return mockIntermediate.GetIntermediateTransactions()
+}
+
+// CleaAccumulatedIntermediateTransactions -
+func CleaAccumulatedIntermediateTransactions(t *testing.T, testContext *vm.VMTestContext) {
```
Comment (SebastianMarian) [inline integrationTests/vm/txsFee/utils/utils.go:132]: typo: Clea instead Clean ? here and in comment

File: process/smartContract/process.go [hunk 21]
```diff
@@ -971,6 +971,13 @@ func (sc *scProcessor) ProcessIfError(
 		return err
 	}
 
+	txType, _ := sc.txTypeHandler.ComputeTransactionType(tx)
+	isCrossShardMoveBalance := txType == process.MoveBalance && check.IfNil(acntSnd)
+	if isCrossShardMoveBalance {
+		// move balance was already consumed in sender shard
+		return nil
```
Comment (SebastianMarian) [inline process/smartContract/process.go:978]: This change will keep the backwards compatibility ?

File: integrationTests/vm/txsFee/multiShard/relayedMoveBalance_test.go (hunk 12 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/multiShard/relayedMoveBalance_test.go:57]: done. test added.

File: integrationTests/vm/txsFee/utils/utils.go (hunk 19 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/utils/utils.go:51]: done.

File: process/smartContract/process.go (hunk 21 above)
Comment (sasurobert) [reply process/smartContract/process.go:978]: We can add a flag, only possibility is sending value to SC cross shard

File: integrationTests/vm/txsFee/multiShard/relayedTxScCalls_test.go (hunk 16 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/multiShard/relayedTxScCalls_test.go:138]: refactored.

File: integrationTests/vm/txsFee/utils/utils.go (hunk 20 above)
Comment (miiu96) [reply integrationTests/vm/txsFee/utils/utils.go:132]: refactored.

File: process/smartContract/process.go (hunk 21 above)
Comment (miiu96) [reply process/smartContract/process.go:978]: done.

**Review by sasurobert — DISMISSED** [review]

**Review by SebastianMarian — DISMISSED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]
System tests passed.

**Pull request merged**